mmlearn/modules/encoders/text.py

```diff
@@ -16,42 +18,123 @@
-        the model are frozen. If it is a float, the first `N` percent of the layers
-        are frozen. If it is a list of integers, the layers at the indices in the
-        list are frozen.
-    freeze_layer_norm : bool, default=True
-        Whether to freeze the layer normalization layers of the model.
-    peft_config : PeftConfig, optional, default=None
-        The configuration from the `peft` library to use to wrap the model
-        for parameter-efficient finetuning.
-    model_config_kwargs : Dict[str, Any], optional, default=None
-        Additional keyword arguments to pass to the model configuration.
-
-    Warns
-    -----
-    UserWarning
-        If both `peft_config` and `freeze_layers` are set. The `peft_config` will
-        override the `freeze_layers` setting.
+    model : PreTrainedModel
+        The model whose layers to freeze.
+    freeze_layers : Union[int, float, List[int], bool]
+        Specification of which layers to freeze.
+    freeze_layer_norm : bool
+        Whether to freeze layer normalization layers.
+    """
+    if isinstance(freeze_layers, bool) and freeze_layers:
+        for name, param in model.named_parameters():
+            param.requires_grad = (
+                (not freeze_layer_norm) if "LayerNorm" in name else False
+            )
+        return
 
+    if isinstance(freeze_layers, (float, int, list)) and model.config.model_type in [
+        "flaubert",
+        "xlm",
+    ]:
+        raise ValueError(
+            f"Freezing individual layers is not supported for {model.config.model_type} "
+            "models. Please use `freeze_layers=False` or `freeze_layers=True`."
+        )
 
-    """
+    # Get list of layers
+    embeddings = model.embeddings
+    encoder = getattr(model, "encoder", None) or getattr(model, "transformer", model)
+    encoder_layers = (
+        getattr(encoder, "layer", None)
+        or getattr(encoder, "layers", None)
+        or getattr(encoder, "block", None)
+    )
+
+    if encoder_layers is None and hasattr(encoder, "albert_layer_groups"):
+        encoder_layers = [
+            layer
+            for group in encoder.albert_layer_groups
+            for layer in group.albert_layers
+        ]
+
+    modules = [embeddings]
+    if encoder_layers is not None and isinstance(encoder_layers, list):
+        modules.extend(encoder_layers)
+
+    if isinstance(freeze_layers, float):
+        freeze_layers = int(freeze_layers * len(modules))
+    if isinstance(freeze_layers, int):
+        freeze_layers = list(range(freeze_layers))
+
+    if isinstance(freeze_layers, list):
+        for idx, module in enumerate(modules):
+            if idx in freeze_layers:
+                for name, param in module.named_parameters():
+                    param.requires_grad = (
+                        (not freeze_layer_norm) if "LayerNorm" in name else False
+                    )
+
+
+class RegressionHead(nn.Module):
```

Comment:
Please reuse [this](https://github.com/VectorInstitute/mmlearn/blob/main/mmlearn/modules/layers/mlp.py) here.

Reply:
Done